##  MNIST 데이터 불러오기

1만 개의 MNIST 훈련용/테스트용 데이터셋을 이전에 저장했던 바이너리 파일에서 불러오겠습니다.

In [2]:
import pickle

mnist_small = pickle.load(open(r'D:/Dev/Python-ML/data/mnist/mnist_small.pickle', 'rb'))

X_train, Y_train, X_test, Y_test = mnist_small

전체 데이터셋을 사용하고 싶다면 아래와 같은 코드를 입력합니다.

In [ ]:
from tensorflow.examples.tutorials.mnist import input_data

mnist = input_data.read_data_sets('D:/Dev/Python-ML/data/mnist/', one_hot=True)

## 데이터 분류 방법

이산적인 값으로 레이블된 데이터를 분류하는 방법에는 여러가지가 있습니다.

- 서포트 벡터 머신
- 랜덤 포레스트
- 로지스틱 회귀
- 신경망
- 컨벌루젼 신경망
- RNN


분류 알고리즘을 사용하는 환경에도 다양한 옵션이 있습니다.

- 파이썬 직접 코딩
- scikit-learn
- Tensorflow
- Keras
- Pytorch

물론 파이썬이 아닌 다른 언어에서 사용 가능한 모듈도 다양합니다.

신경망을 직접 코딩하는 것은 그리 쉽지는 않지만 그리 어렵지도 않습니다. 신경망의 퍼셉트론과 훈련 알고리즘을 이애해야하면 코딩할 수 있습니다.  컨번루젼 신경망을 코딩하려면 컨벌루젼 층의 다차원 배열의 변환에 익숙해야 합니다.

사이킷런에서는 Depp neural network 를 제대로 구현할 수는 없고, 다층 퍼센트론만 구현할 수 있습니다.

텐서플로우는 배우기 쉬운 머신러닝 언어는 아닙니다.  하지만 신경망 이론을 이해하고 있다면 다른 사람의 코드를 보고 쉽게 응용할 수 있습니다.

Keras, Pytorch 에서는 더 간편하게 신경망을 구현할 수 있습니다.

신경망을 이해하고 Tensorflow, Keras, Pytorch 중에서 하나를 다룰 줄 알면 (텐서플로우만은 쉽다고 까지는 말을 못하지만) 나머지는 그리 어렵지 않게 배울 수 있습니다.

머신러닝을 처음 배우는 단계이고 현업에서 사용하는 것이 아니라면 신경망 구성은 keras 로 익히고 나머지 머신러닝 시스템에는 사이킷런을 사용하는 것을 권합니다.

먼저 여기서는 신경망 표현이 간단한 keras 를 사용해서 신경망 구성에 대해서 설명드리겠습니다.

keras 는 pip 로 인스톨 합니다.   
\>>> pip install keras

In [30]:
import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten, Activation
from keras.layers import Conv2D, MaxPooling2D
from keras.optimizers import SGD
from keras import metrics

임포트하는 모듈도 많은데 간단하게 설명하겠습니다.

Sequential 은 신경망 레이어를 담는 그릇입니다.  Sequential 의 인스턴슬르 만들고 레이어를 연속해서 추가합니다.

추가하는 레이어가 많을 수록 깊은 즉 deep neural network 가 됩니다.


In [13]:
model = Sequential()

model.add(Dense(units=64, input_dim=784))
model.add(Activation('relu'))
model.add(Dense(units=10))
model.add(Activation('softmax'))

먼저 Dense 레이어를 추가했습니다. 신경망의 하나의 은닉층을 표현합니다.  

MNIST 의 입력 데이터의 크기인 784 를 inpu_dim 에 넣어줍니다.

첫 번째 은닉층의 유닛(노드)는 64개 입니다.  입력 유닛의 크기인 784에 비해서 상당히 작은 값이지만 우선 이렇게 훈련시켜 보겠습니다.

그 다음 층에 신호를 전달하기 전에 신호에 ReLU 함수를 적용시킵니다.    신경망에서는 활성함수로 sigmoid 대신에 ReLU 함수를 사용합니다.

이 경우에는 층이 깊지 않으므로 sigmoid 함수를 사용해도 전혀 문제가 없습니다.

그 다음 은닉층은 없고 바로 출력층입니다.  출력에서 예측해야하는 종류는 0부터 9까지의 10 종류의 숫자 입니다.  출력층의 유닛은 10으로 정합니다.

출력층에서 나온 결과를 softmax 함수를 통과시키면 각 레이블별 확률로 해석할 수 있는 값이 계산됩니다.


그 값을 중에서 가장 큰 값을 가지는 노드에 해당하는 레이블이 예측값이 됩니다.

In [14]:
model.compile(loss='categorical_crossentropy',
              optimizer='sgd',
              metrics=['accuracy'])

이 렇게 만든 모델에 대해서 추가 옵션을 정해서 컴파일을 합니다.

loss 함수는 당연히 cross entropy 를 사용하고

optimizer 에서는 스토캐스틱 gradient descent 방법을 사용합니다.

출력 결과물은 'accuracy' 정확도만으로 설정했습니다.

아래는 실제 데이터로 훈련시키는 과정입니다.

전체 데이터에 대해서 총 10회 반복한고 한 번에 훈련시키는 배치의 크기는 100 입니다

In [15]:
model.fit(X_train, Y_train, epochs=10, batch_size=100)

Epoch 1/10
10000/10000 [==============================] - ETA: 0s - loss: 1.9790 - acc: 0.387 - 1s - loss: 1.9722 - acc: 0.3912     
Epoch 2/10
10000/10000 [==============================] - 1s - loss: 1.3584 - acc: 0.7125     
Epoch 3/10
10000/10000 [==============================] - 1s - loss: 0.9999 - acc: 0.7848     
Epoch 4/10
10000/10000 [==============================] - 1s - loss: 0.8091 - acc: 0.8205     
Epoch 5/10
10000/10000 [==============================] - 1s - loss: 0.6963 - acc: 0.8390     
Epoch 6/10
10000/10000 [==============================] - 1s - loss: 0.6226 - acc: 0.8514     
Epoch 7/10
10000/10000 [==============================] - 1s - loss: 0.5703 - acc: 0.8594     
Epoch 8/10
10000/10000 [==============================] - 1s - loss: 0.5315 - acc: 0.8672     
Epoch 9/10
10000/10000 [==============================] - 1s - loss: 0.5006 - acc: 0.8727     
Epoch 10/10
10000/10000 [==============================] - 1s - loss: 0.4762 - acc: 0.8788     


훈련이 진행될 수록 loss (또는 cost) 가 줄어드는 것을 확인할 수 있습니다.  정확도는 점점 좋아져서  훈련데이터의 정확도가 87.88% 입니다.

모델 성능에 대한 평가에는 테스트셋을 사용합니다.

In [16]:
loss_and_metrics = model.evaluate(X_test, Y_test, batch_size=128)

 9216/10000 [==========================>...] - ETA: 0s

In [17]:
loss_and_metrics

[0.45433075084686281, 0.88219999999999998]

테스트셋에 대해서 모델을 평가하니 정확도가 88.21% 가 나왔습니다.

이전의 SVM 에서는 92.36% 가 나왔으니 지금은 형편없는 신경망입니다.

신경망을 더 깊고 넖게 만들어서 훈련시켜보겠습니다.

In [18]:
model = Sequential()

# 입력층 + 첫 번쨰 은닉층
model.add(Dense(units=512, input_dim=784))
model.add(Activation('relu'))

# 은닉층 
model.add(Dense(units=512))
model.add(Activation('relu'))
model.add(Dense(units=256))
model.add(Activation('relu'))
# 오버피팅을 방지하기 위한 층
# model.add(Dropout(0.5))


# 출력층
model.add(Dense(units=10))
model.add(Activation('softmax'))

은닉층을 추가해주는 방법은 unit 의 갯수를 정해서 model.add(Dense(units=512)) 와 같이 추가한 후에 

활성화층인  model.add(Activation('relu')) 를 추가해주면 됩니다.  

10 개 정도를 사용할 수도 있고 100 개 이상을 사용할 수도 있습니다.

오버피팅을 방지하기 위해서 Drop out 층을 넣을 수도 있습니다.

이번에는 optimizer 에 추가 파라미터도 지정해봤습니다.  learning_rate 는 0.01 이고 momentum 은 0.9입니다.

In [25]:
model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.SGD(lr=0.01, momentum=0.9, nesterov=True),
              metrics=[metrics.mae, metrics.categorical_accuracy])

In [26]:
model.fit(X_train, Y_train, epochs=10, batch_size=100)

Epoch 1/10
10000/10000 [==============================] - 8s - loss: 0.0496 - mean_absolute_error: 0.0069 - categorical_accuracy: 0.9889     
Epoch 2/10
10000/10000 [==============================] - 7s - loss: 0.0413 - mean_absolute_error: 0.0059 - categorical_accuracy: 0.9913     
Epoch 3/10
10000/10000 [==============================] - 8s - loss: 0.0329 - mean_absolute_error: 0.0050 - categorical_accuracy: 0.9942     
Epoch 4/10
10000/10000 [==============================] - 8s - loss: 0.0286 - mean_absolute_error: 0.0045 - categorical_accuracy: 0.9953     
Epoch 5/10
10000/10000 [==============================] - 8s - loss: 0.0209 - mean_absolute_error: 0.0034 - categorical_accuracy: 0.9972     
Epoch 6/10
10000/10000 [==============================] - 7s - loss: 0.0173 - mean_absolute_error: 0.0030 - categorical_accuracy: 0.9983     
Epoch 7/10
10000/10000 [==============================] - 8s - loss: 0.0143 - mean_absolute_error: 0.0025 - categorical_accuracy: 0.9989     
Epoch 

In [27]:
loss_and_metrics = model.evaluate(X_test, Y_test, batch_size=128)


 9984/10000 [============================>.] - ETA: 0s

In [29]:
loss_and_metrics

[0.16518993893722073, 0.010180272687831893, 0.95830000000000004]

조금 더 갚은 신경망을 훈련시킨 결과 테스트셋에서 95.83% 의 정확도를 얻었습니다.

신경망의 파라미터와 옵션을 조정해서 실습을 해볼 수 있습니다.